rust_src/src/base64.rs

```diff
@@ -68,32 +68,37 @@ pub extern "C" fn base64_encode_1(
 #[no_mangle]
 pub extern "C" fn base64_decode_1(
     from: *const c_char,
-    to: *mut c_char,
     length: ptrdiff_t,
+    to: *mut c_char,
+    out_length: ptrdiff_t,
     multibyte: bool,
     nchars_return: *mut ptrdiff_t,
 ) -> ptrdiff_t {
     let encoded = unsafe { slice::from_raw_parts(from as *const u8, length as usize) };
+    let decoded = unsafe { slice::from_raw_parts_mut(to as *mut u8, out_length as usize) };
 
     // Use the MIME config to allow embedded newlines.
-    if let Ok(decoded) = base64_crate::decode_config(encoded, base64_crate::MIME) {
+    if let Ok(decoded_length) =
+        base64_crate::decode_config_slice(encoded, base64_crate::MIME, decoded)
+    {
         if !nchars_return.is_null() {
             unsafe {
-                *nchars_return = decoded.len() as ptrdiff_t;
+                *nchars_return = decoded_length as ptrdiff_t;
             }
         }
         if multibyte {
             // Decode non-ASCII bytes into UTF-8 pairs.
-            let s: String = decoded.iter().map(|&byte| byte as char).collect();
+            let s: String = decoded[..decoded_length]
+                .iter()
+                .map(|&byte| byte as char)
+                .collect();
+            let s_len = s.len();
             unsafe {
-                ptr::copy_nonoverlapping(s.as_ptr(), to as *mut u8, s.len());
+                ptr::copy_nonoverlapping(s.as_ptr(), to as *mut u8, s_len);
```

Comment:
Can't we skip this copy? We should be able to decode directly into the `to` buffer...

Reply:
I had weird failures when I tried so I put it back. I will look at it again.

Reply:
We do decode directly into `to`. It is named `decoded` since I had to make it a slice in Rust.

The issue is we have to handle the Emacs multibyte transcoding. The else hard does exactly as you suggest and simply uses `to`.